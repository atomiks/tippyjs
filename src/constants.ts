export const PASSIVE = {passive: true};

export const ROUND_ARROW =
  '<svg viewBox="0 0 18 7" xmlns="http://www.w3.org/2000/svg"><path d="M0 7s2.021-.015 5.253-4.218C6.584 1.051 7.797.007 9 0c1.203-.007 2.416 1.035 3.761 2.782C16.012 7.005 18 7 18 7H0z"/></svg>';

export const IOS_CLASS = `__NAMESPACE_PREFIX__-iOS`;
export const POPPER_CLASS = `__NAMESPACE_PREFIX__-popper`;
export const TOOLTIP_CLASS = `__NAMESPACE_PREFIX__-tooltip`;
export const CONTENT_CLASS = `__NAMESPACE_PREFIX__-content`;
export const BACKDROP_CLASS = `__NAMESPACE_PREFIX__-backdrop`;
export const ARROW_CLASS = `__NAMESPACE_PREFIX__-arrow`;
export const SVG_ARROW_CLASS = `__NAMESPACE_PREFIX__-svg-arrow`;

export const POPPER_SELECTOR = `.${POPPER_CLASS}`;
export const TOOLTIP_SELECTOR = `.${TOOLTIP_CLASS}`;
export const CONTENT_SELECTOR = `.${CONTENT_CLASS}`;
export const BACKDROP_SELECTOR = `.${BACKDROP_CLASS}`;
export const ARROW_SELECTOR = `.${ARROW_CLASS}`;
export const SVG_ARROW_SELECTOR = `.${SVG_ARROW_CLASS}`;
