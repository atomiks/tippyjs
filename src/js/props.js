import { getDataAttributeOptions } from './reference'
import { hasOwnProperty } from './utils'

/**
 * Evaluates options
 * @param {Element} reference
 * @param {Object} options
 * @return {Object}
 */
export function evaluateProps(reference, props) {
  const out = {
    ...props,
    ...(props.performance ? {} : getDataAttributeOptions(reference))
  }

  if (out.arrow) {
    out.animateFill = false
  }

  if (typeof out.appendTo === 'function') {
    out.appendTo = props.appendTo(reference)
  }

  if (typeof out.content === 'function') {
    out.content = props.content(reference)
  }

  return out
}

/**
 * Validates an object of options with the valid default props object
 * @param {Object} options
 * @param {Object} defaults
 */
export function validateOptions(options = {}, defaults) {
  Object.keys(options).forEach(option => {
    if (!hasOwnProperty(defaults, option)) {
      throw new Error(`[tippy]: \`${option}\` is not a valid option`)
    }
  })
}
