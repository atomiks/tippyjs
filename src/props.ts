import {DefaultProps, Plugin, Props, ReferenceElement, Tippy} from './types';
import {
  hasOwnProperty,
  removeProperties,
  invokeWithArgsOrReturn,
} from './utils';
import {warnWhen} from './validation';

const pluginProps = {
  animateFill: false,
  followCursor: false,
  inlinePositioning: false,
  sticky: false,
};

const renderProps = {
  allowHTML: false,
  animation: 'fade',
  arrow: true,
  content: '',
  inertia: false,
  maxWidth: 350,
  role: 'tooltip',
  theme: '',
  zIndex: 9999,
};

export const defaultProps: DefaultProps = {
  appendTo: () => document.body,
  aria: {
    content: 'auto',
    expanded: 'auto',
  },
  delay: 0,
  duration: [300, 250],
  getReferenceClientRect: null,
  hideOnClick: true,
  ignoreAttributes: false,
  interactive: false,
  interactiveBorder: 2,
  interactiveDebounce: 0,
  moveTransition: '',
  offset: [0, 10],
  onAfterUpdate() {},
  onBeforeUpdate() {},
  onCreate() {},
  onDestroy() {},
  onHidden() {},
  onHide() {},
  onMount() {},
  onShow() {},
  onShown() {},
  onTrigger() {},
  onUntrigger() {},
  onClickOutside() {},
  placement: 'top',
  plugins: [],
  popperOptions: {},
  render: null,
  showOnCreate: false,
  touch: true,
  trigger: 'mouseenter focus',
  triggerTarget: null,
  ...pluginProps,
  ...renderProps,
};

const defaultKeys = Object.keys(defaultProps);

export const setDefaultProps: Tippy['setDefaultProps'] = (partialProps) => {
  /* istanbul ignore else */
  if (__DEV__) {
    validateProps(partialProps, []);
  }

  const keys = Object.keys(partialProps) as Array<keyof DefaultProps>;
  keys.forEach((key) => {
    (defaultProps as any)[key] = partialProps[key];
  });
};

export function getExtendedPassedProps(
  passedProps: Partial<Props> & Record<string, unknown>
): Partial<Props> {
  const plugins = passedProps.plugins || [];
  const pluginProps = plugins.reduce<Record<string, unknown>>((acc, plugin) => {
    const {name, defaultValue} = plugin;

    if (name) {
      acc[name] =
        passedProps[name] !== undefined ? passedProps[name] : defaultValue;
    }

    return acc;
  }, {});

  return {
    ...passedProps,
    ...pluginProps,
  };
}

export function getDataAttributeProps(
  reference: ReferenceElement,
  plugins: Plugin[]
): Record<string, unknown> {
  const propKeys = plugins
    ? Object.keys(getExtendedPassedProps({...defaultProps, plugins}))
    : defaultKeys;

  const props = propKeys.reduce(
    (acc: Partial<Props> & Record<string, unknown>, key) => {
      const valueAsString = (
        reference.getAttribute(`data-tippy-${key}`) || ''
      ).trim();

      if (!valueAsString) {
        return acc;
      }

      if (key === 'content') {
        acc[key] = valueAsString;
      } else {
        try {
          acc[key] = JSON.parse(valueAsString);
        } catch (e) {
          acc[key] = valueAsString;
        }
      }

      return acc;
    },
    {}
  );

  return props;
}

export function evaluateProps(
  reference: ReferenceElement,
  props: Props
): Props {
  const out = {
    ...props,
    content: invokeWithArgsOrReturn(props.content, [reference]),
    ...(props.ignoreAttributes
      ? {}
      : getDataAttributeProps(reference, props.plugins)),
  };

  out.aria = {
    ...defaultProps.aria,
    ...out.aria,
  };

  out.aria = {
    expanded:
      out.aria.expanded === 'auto' ? props.interactive : out.aria.expanded,
    content:
      out.aria.content === 'auto'
        ? props.interactive
          ? null
          : 'describedby'
        : out.aria.content,
  };

  return out;
}

export function validateProps(
  partialProps: Partial<Props> = {},
  plugins: Plugin[] = []
): void {
  const keys = Object.keys(partialProps) as Array<keyof Props>;
  keys.forEach((prop) => {
    const nonPluginProps = removeProperties(
      defaultProps,
      Object.keys(pluginProps)
    );

    let didPassUnknownProp = !hasOwnProperty(nonPluginProps, prop);

    // Check if the prop exists in `plugins`
    if (didPassUnknownProp) {
      didPassUnknownProp =
        plugins.filter((plugin) => plugin.name === prop).length === 0;
    }

    warnWhen(
      didPassUnknownProp,
      [
        `\`${prop}\``,
        "is not a valid prop. You may have spelled it incorrectly, or if it's",
        'a plugin, forgot to pass it in an array as props.plugins.',
        '\n\n',
        'All props: https://atomiks.github.io/tippyjs/v6/all-props/\n',
        'Plugins: https://atomiks.github.io/tippyjs/v6/plugins/',
      ].join(' ')
    );
  });
}
