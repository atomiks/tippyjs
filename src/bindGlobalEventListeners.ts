import {PASSIVE} from './constants';
import {isReferenceElement} from './utils';

export const currentInput = {isTouch: false};
let lastMouseMoveTime = 0;

/**
 * When a `touchstart` event is fired, it's assumed the user is using touch
 * input. We'll bind a `mousemove` event listener to listen for mouse input in
 * the future. This way, the `isTouch` property is fully dynamic and will handle
 * hybrid devices that use a mix of touch + mouse input.
 */
export function onDocumentTouchStart(): void {
  if (currentInput.isTouch) {
    return;
  }

  currentInput.isTouch = true;

  if (window.performance) {
    document.addEventListener('mousemove', onDocumentMouseMove);
  }
}

/**
 * When two `mousemove` event are fired consecutively within 20ms, it's assumed
 * the user is using mouse input again. `mousemove` can fire on touch devices as
 * well, but very rarely that quickly.
 */
export function onDocumentMouseMove(): void {
  const now = performance.now();

  if (now - lastMouseMoveTime < 20) {
    currentInput.isTouch = false;

    document.removeEventListener('mousemove', onDocumentMouseMove);
  }

  lastMouseMoveTime = now;
}

/**
 * When an element is in focus and has a tippy, leaving the tab/window and
 * returning causes it to show again. For mouse users this is unexpected, but
 * for keyboard use it makes sense.
 * TODO: find a better technique to solve this problem
 */
export function onWindowBlur(): void {
  const activeElement = document.activeElement as HTMLElement | null;

  if (isReferenceElement(activeElement)) {
    const instance = activeElement._tippy!;

    if (activeElement.blur && !instance.state.isVisible) {
      activeElement.blur();
    }
  }
}

/**
 * Adds the needed global event listeners
 */
export default function bindGlobalEventListeners(): void {
  document.addEventListener('touchstart', onDocumentTouchStart, {
    ...PASSIVE,
    capture: true,
  });
  window.addEventListener('blur', onWindowBlur);
}
