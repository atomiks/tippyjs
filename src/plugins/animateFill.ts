import {AnimateFill} from '../types';
import {BACKDROP_CLASS} from '../constants';
import {div, setVisibilityState} from '../utils';
import {isUCBrowser} from '../browser';
import {warnWhen} from '../validation';

const animateFill: AnimateFill = {
  name: 'animateFill',
  defaultValue: false,
  fn(instance) {
    const {tooltip, content} = instance.popperChildren;

    const backdrop =
      instance.props.animateFill && !isUCBrowser
        ? createBackdropElement()
        : null;

    function addBackdropToPopperChildren(): void {
      instance.popperChildren.backdrop = backdrop;
    }

    return {
      onCreate(): void {
        if (backdrop) {
          addBackdropToPopperChildren();

          tooltip.insertBefore(backdrop, tooltip.firstElementChild!);
          tooltip.setAttribute('data-animatefill', '');
          tooltip.style.overflow = 'hidden';

          instance.setProps({animation: 'shift-away', arrow: false});
        }
      },
      onMount(): void {
        if (backdrop) {
          const {transitionDuration} = tooltip.style;
          const duration = Number(transitionDuration.replace('ms', ''));

          // The content should fade in after the backdrop has mostly filled the
          // tooltip element. `clip-path` is the other alternative but is not
          // well-supported and is buggy on some devices.
          content.style.transitionDelay = `${Math.round(duration / 10)}ms`;

          backdrop.style.transitionDuration = transitionDuration;
          setVisibilityState([backdrop], 'visible');

          // Warn if the stylesheets are not loaded
          if (__DEV__) {
            warnWhen(
              getComputedStyle(backdrop).position !== 'absolute',
              `The \`tippy.js/dist/backdrop.css\` stylesheet has not been
              imported!
              
              The \`animateFill\` plugin requires this stylesheet to work.`,
            );

            warnWhen(
              getComputedStyle(tooltip).transform === 'none',
              `The \`tippy.js/animations/shift-away.css\` stylesheet has not
              been imported!
              
              The \`animateFill\` plugin requires this stylesheet to work.`,
            );
          }
        }
      },
      onShow(): void {
        if (backdrop) {
          backdrop.style.transitionDuration = '0ms';
        }
      },
      onHide(): void {
        if (backdrop) {
          setVisibilityState([backdrop], 'hidden');
        }
      },
      onAfterUpdate(): void {
        // With this type of prop, it's highly unlikely it will be changed
        // dynamically. We'll leave out the diff/update logic it to save bytes.

        // `popperChildren` is assigned a new object onAfterUpdate
        addBackdropToPopperChildren();
      },
    };
  },
};

export default animateFill;

function createBackdropElement(): HTMLDivElement {
  const backdrop = div();
  backdrop.className = BACKDROP_CLASS;
  setVisibilityState([backdrop], 'hidden');
  return backdrop;
}
