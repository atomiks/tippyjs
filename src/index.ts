import {version} from '../package.json';
import {defaultProps} from './props';
import createTippy from './createTippy';
import bindGlobalEventListeners, {
  currentInput,
} from './bindGlobalEventListeners';
import {
  arrayFrom,
  getArrayOfElements,
  isReferenceElement,
  isElement,
} from './utils';
import {warnWhen, validateTargets, validateProps} from './validation';
import {POPPER_SELECTOR} from './constants';
import {
  Props,
  Instance,
  Targets,
  PopperElement,
  HideAllOptions,
  Plugin,
  Tippy,
} from './types';

/**
 * Exported module
 */
function tippy(
  targets: Targets,
  optionalProps?: Partial<Props>,
  plugins: Plugin[] = [],
): Instance | Instance[] {
  if (__DEV__) {
    validateTargets(targets);
    validateProps(optionalProps, plugins);
  }

  bindGlobalEventListeners();

  const props: Props = {...defaultProps, ...optionalProps};

  const elements = getArrayOfElements(targets);

  if (__DEV__) {
    const isSingleContentElement = isElement(props.content);
    const isMoreThanOneReferenceElement = elements.length > 1;
    warnWhen(
      isSingleContentElement && isMoreThanOneReferenceElement,
      `tippy() was passed an Element as the \`content\` prop, but more than one
      tippy instance was created by this invocation. This means the content
      element will only be appended to the last tippy instance.
      
      Instead, pass the .innerHTML of the element, or use a function that
      returns a cloned version of the element instead.
      
      1) content: () => element.cloneNode(true)
      2) content: element.innerHTML`,
    );
  }

  const instances = elements.reduce<Instance[]>(
    (acc, reference): Instance[] => {
      const instance = reference && createTippy(reference, props, plugins);

      if (instance) {
        acc.push(instance);
      }

      return acc;
    },
    [],
  );

  return isElement(targets) ? instances[0] : instances;
}

tippy.version = version;
tippy.defaultProps = defaultProps;
tippy.setDefaultProps = setDefaultProps;
tippy.currentInput = currentInput;

/**
 * Mutates the defaultProps object by setting the props specified
 */
function setDefaultProps(partialProps: Partial<Props>): void {
  if (__DEV__) {
    validateProps(partialProps, []);
  }

  Object.keys(partialProps).forEach((key): void => {
    defaultProps[key] = partialProps[key];
  });
}

/**
 * Returns a proxy wrapper function that passes the plugins
 */
export function createTippyWithPlugins(outerPlugins: Plugin[]): Tippy {
  const fn = (
    targets: Targets,
    optionalProps?: Partial<Props>,
    innerPlugins: Plugin[] = [],
  ): Instance | Instance[] =>
    tippy(targets, optionalProps, [...outerPlugins, ...innerPlugins]);

  fn.version = version;
  fn.defaultProps = defaultProps;
  fn.setDefaultProps = setDefaultProps;
  fn.currentInput = currentInput;

  return fn;
}

/**
 * Hides all visible poppers on the document
 */
export function hideAll({
  exclude: excludedReferenceOrInstance,
  duration,
}: HideAllOptions = {}): void {
  arrayFrom(document.querySelectorAll(POPPER_SELECTOR)).forEach(
    (popper: PopperElement): void => {
      const instance = popper._tippy;

      if (instance) {
        let isExcluded = false;
        if (excludedReferenceOrInstance) {
          isExcluded = isReferenceElement(excludedReferenceOrInstance)
            ? instance.reference === excludedReferenceOrInstance
            : popper === excludedReferenceOrInstance.popper;
        }

        if (!isExcluded) {
          instance.hide(duration);
        }
      }
    },
  );
}

export default tippy;
