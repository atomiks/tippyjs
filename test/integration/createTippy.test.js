import {fireEvent} from '@testing-library/dom';
import {
  h,
  cleanDocumentBody,
  setTestDefaultProps,
  enableTouchEnvironment,
  disableTouchEnvironment,
} from '../utils';

import {defaultProps} from '../../src/props';
import createTippy, {mountedInstances} from '../../src/createTippy';
import {POPPER_SELECTOR, IOS_CLASS} from '../../src/constants';
import animateFill from '../../src/plugins/animateFill';

jest.useFakeTimers();
setTestDefaultProps();

let instance;

afterEach(() => {
  if (instance) {
    instance.destroy();
  }
  cleanDocumentBody();
});

describe('createTippy', () => {
  it('returns `null` if the reference already has a tippy instance', () => {
    const reference = h();
    reference._tippy = true;

    expect(createTippy(reference, defaultProps)).toBe(null);
  });

  it('returns the instance with expected properties', () => {
    instance = createTippy(h(), defaultProps);

    expect(instance.id).toBeDefined();
    expect(instance.reference).toBeDefined();
    expect(instance.popper).toBeDefined();
    expect(instance.popperInstance).toBeDefined();
    expect(instance.popperChildren).toBeDefined();
    expect(instance.popperInstance).toBeDefined();
    expect(instance.props).toBeDefined();
    expect(instance.state).toBeDefined();
    expect(instance.clearDelayTimeouts).toBeDefined();
    expect(instance.setProps).toBeDefined();
    expect(instance.setContent).toBeDefined();
    expect(instance.show).toBeDefined();
    expect(instance.hide).toBeDefined();
    expect(instance.enable).toBeDefined();
    expect(instance.disable).toBeDefined();
    expect(instance.destroy).toBeDefined();
  });

  it('increments the `id` on each call with valid arguments', () => {
    const instances = [
      createTippy(h(), defaultProps),
      createTippy(h(), defaultProps),
      createTippy(h(), defaultProps),
    ];

    expect(instances[0].id).toBe(instances[1].id - 1);
    expect(instances[1].id).toBe(instances[2].id - 1);
  });

  it('adds correct listeners to the reference element based on `trigger`', () => {
    instance = createTippy(h(), {
      ...defaultProps,
      trigger: 'mouseenter focus click',
    });

    fireEvent.mouseEnter(instance.reference);
    expect(instance.state.isVisible).toBe(true);

    fireEvent.mouseLeave(instance.reference);
    expect(instance.state.isVisible).toBe(false);

    fireEvent.focus(instance.reference);
    expect(instance.state.isVisible).toBe(true);

    fireEvent.blur(instance.reference);
    expect(instance.state.isVisible).toBe(false);

    fireEvent.click(instance.reference);
    expect(instance.state.isVisible).toBe(true);

    fireEvent.click(instance.reference);
    expect(instance.state.isVisible).toBe(false);
  });

  it('extends `instance.props` with plugin props', () => {
    instance = createTippy(h(), {...defaultProps, plugins: [animateFill]});

    expect(instance.props.animateFill).toBe(animateFill.defaultValue);
  });

  it('`instance.plugins` does not contain duplicate plugins', () => {
    instance = createTippy(h(), {
      ...defaultProps,
      plugins: [animateFill, animateFill],
    });

    expect(instance.plugins).toEqual([animateFill]);
  });
});

describe('instance.destroy()', () => {
  it('sets state.isDestroyed to `true`', () => {
    instance = createTippy(h(), defaultProps);
    instance.destroy();

    expect(instance.state.isDestroyed).toBe(true);
  });

  it('deletes the `_tippy` property from the reference', () => {
    const ref = h();
    instance = createTippy(ref, defaultProps);

    expect('_tippy' in ref).toBe(true);

    instance.destroy();

    expect('_tippy' in ref).toBe(false);
  });

  it('removes listeners from the reference', () => {
    const ref = h();

    instance = createTippy(ref, {
      ...defaultProps,
      trigger: 'mouseenter',
    });

    instance.destroy();
    fireEvent.mouseEnter(ref);

    expect(instance.state.isVisible).toBe(false);
  });

  it('does nothing if the instance is already destroyed', () => {
    const ref = h();

    instance = createTippy(ref, defaultProps);
    instance.state.isDestroyed = true;
    instance.destroy();

    expect(ref._tippy).toBeDefined();
  });

  it('still unmounts the tippy if the instance is disabled', () => {
    instance = createTippy(h(), defaultProps);

    instance.show();
    instance.disable();
    instance.destroy();

    expect(instance.state.isMounted).toBe(false);
  });

  it('still unmounts the tippy if the instance is disabled', () => {
    instance = createTippy(h(), defaultProps);

    instance.show();

    jest.runAllTimers();

    instance.disable();
    instance.destroy();

    expect(instance.state.isMounted).toBe(false);
  });

  it('clears pending timeouts', () => {
    instance = createTippy(h(), {...defaultProps, delay: 500});

    // show() will warn about memory leak
    const spy = jest.spyOn(console, 'warn');

    fireEvent.mouseEnter(instance.reference);

    instance.destroy();

    jest.advanceTimersByTime(500);

    expect(spy).not.toHaveBeenCalled();
  });
});

describe('instance.show()', () => {
  it('changes state.isVisible to `true`', () => {
    instance = createTippy(h(), defaultProps);
    instance.show();

    expect(instance.state.isVisible).toBe(true);
  });

  it('mounts the popper to the DOM', () => {
    instance = createTippy(h(), defaultProps);
    instance.show();
    jest.runAllTimers();

    expect(document.querySelector(POPPER_SELECTOR)).toBe(instance.popper);
  });

  it('overrides instance.props.duration if supplied an argument', () => {
    instance = createTippy(h(), {
      ...defaultProps,
      duration: 100,
    });

    instance.show(10);
    jest.runAllTimers();

    expect(instance.popperChildren.tooltip.style.transitionDuration).toBe(
      '10ms',
    );
  });

  it('does not show tooltip if the reference has a `disabled` attribute', () => {
    const ref = h();

    ref.setAttribute('disabled', 'disabled');
    instance = createTippy(ref, defaultProps);
    instance.show();

    expect(instance.state.isVisible).toBe(false);
  });

  it('bails out if already visible', () => {
    const spy = jest.fn();

    instance = createTippy(h(), {...defaultProps, onShow: spy});
    instance.show();
    instance.show();

    expect(spy).toHaveBeenCalledTimes(1);
  });

  it('sets elements to duration 0 before mount, unless already mounted', () => {
    instance = createTippy(h(), {
      ...defaultProps,
      updateDuration: 99,
      duration: 99,
    });

    const {popper} = instance;
    const {tooltip, content} = instance.popperChildren;

    instance.show();

    expect(popper.style.transitionDuration).toBe('0ms');
    expect(tooltip.style.transitionDuration).toBe('0ms');
    expect(content.style.transitionDuration).toBe('0ms');

    jest.runAllTimers();

    expect(popper.style.transitionDuration).toBe('99ms');
    expect(tooltip.style.transitionDuration).toBe('99ms');
    expect(content.style.transitionDuration).toBe('99ms');

    instance.hide(99);
    instance.show();

    expect(popper.style.transitionDuration).toBe('99ms');
    expect(tooltip.style.transitionDuration).toBe('99ms');
    expect(content.style.transitionDuration).toBe('99ms');
  });
});

describe('instance.hide()', () => {
  it('changes state.isVisible to false', () => {
    instance = createTippy(h(), defaultProps);
    instance.hide();

    expect(instance.state.isVisible).toBe(false);
  });

  it('removes the popper element from the DOM after hiding', () => {
    instance = createTippy(h(), {
      ...defaultProps,
    });

    instance.show(0);
    jest.runAllTimers();

    expect(document.querySelector(POPPER_SELECTOR)).toBe(instance.popper);

    instance.hide(0);
    jest.runAllTimers();

    expect(document.querySelector(POPPER_SELECTOR)).toBeNull();
  });

  it('overrides instance.props.duration if supplied an argument', () => {
    instance = createTippy(h(), {
      ...defaultProps,
      duration: 100,
    });

    instance.show(0);
    jest.runAllTimers();
    instance.hide(9);

    expect(instance.popperChildren.tooltip.style.transitionDuration).toBe(
      '9ms',
    );
  });

  it('bails out if already hidden', () => {
    const spy = jest.fn();

    instance = createTippy(h(), {...defaultProps, onHide: spy});
    instance.hide();
    instance.hide();

    expect(spy).toHaveBeenCalledTimes(0);

    instance.show();
    instance.hide();
    instance.hide();

    expect(spy).toHaveBeenCalledTimes(1);
  });
});

describe('instance.enable()', () => {
  it('sets state.isEnabled to `true`', () => {
    instance = createTippy(h(), defaultProps);
    instance.enable();

    expect(instance.state.isEnabled).toBe(true);
  });

  it('allows a tippy to be shown', () => {
    instance = createTippy(h(), defaultProps);
    instance.enable();
    instance.show();

    expect(instance.state.isVisible).toBe(true);
  });
});

describe('instance.disable()', () => {
  it('sets state.isEnabled to `false`', () => {
    instance = createTippy(h(), defaultProps);
    instance.disable();

    expect(instance.state.isEnabled).toBe(false);
  });

  it('disallows a tippy to be shown', () => {
    instance = createTippy(h(), defaultProps);
    instance.disable();
    instance.show();

    expect(instance.state.isVisible).toBe(false);
  });

  it('hides the instance if visible', () => {
    instance = createTippy(h(), defaultProps);
    instance.show();
    instance.disable();

    expect(instance.state.isVisible).toBe(false);
  });
});

describe('instance.setProps()', () => {
  it('sets the new props by merging them with the current instance', () => {
    instance = createTippy(h(), defaultProps);

    expect(instance.props.arrow).toBe(defaultProps.arrow);
    expect(instance.props.duration).toBe(defaultProps.duration);

    instance.setProps({arrow: !defaultProps.arrow, duration: 82});

    expect(instance.props.arrow).toBe(!defaultProps.arrow);
    expect(instance.props.duration).toBe(82);
  });

  it('redraws the tooltip by creating a new popper element', () => {
    instance = createTippy(h(), defaultProps);

    expect(
      instance.popper.querySelector('.__NAMESPACE_PREFIX__-arrow'),
    ).not.toBeNull();

    instance.setProps({arrow: false});

    expect(
      instance.popper.querySelector('.__NAMESPACE_PREFIX__-arrow'),
    ).toBeNull();
  });

  it('popperChildren property is updated to reflect the new popper element', () => {
    instance = createTippy(h(), defaultProps);

    expect(instance.popperChildren.arrow).not.toBeNull();

    instance.setProps({arrow: false});

    expect(instance.popperChildren.arrow).toBeNull();
  });

  it('popperInstance popper is updated to the new popper', () => {
    instance = createTippy(h(), {
      ...defaultProps,
      lazy: false,
    });
    instance.setProps({arrow: true});
    expect(instance.popperInstance.popper).toBe(instance.popper);
  });

  it('popper._tippy is defined with the correct instance', () => {
    instance = createTippy(h(), defaultProps);
    instance.setProps({arrow: true});
    expect(instance.popper._tippy).toBe(instance);
  });

  it('changing `trigger` or `touchHold` changes listeners', () => {
    const ref = h();
    instance = createTippy(ref, defaultProps);

    instance.setProps({trigger: 'click'});
    fireEvent.mouseEnter(ref);

    expect(instance.state.isVisible).toBe(false);

    fireEvent.click(ref);
    expect(instance.state.isVisible).toBe(true);
  });

  it('avoids creating a new popperInstance if new props are identical', () => {
    instance = createTippy(h(), defaultProps);
    instance.show();
    const previousPopperInstance = instance.popperInstance;
    instance.setProps(defaultProps);

    expect(instance.popperInstance).toBe(previousPopperInstance);
  });

  it('creates a new popperInstance if one of the props has changed', () => {
    instance = createTippy(h(), defaultProps);
    instance.show();
    const previousPopperInstance = instance.popperInstance;
    instance.setProps({...defaultProps, placement: 'bottom'});

    expect(instance.popperInstance).not.toBe(previousPopperInstance);
  });

  it('does nothing if the instance is destroyed', () => {
    instance = createTippy(h(), defaultProps);

    instance.destroy();
    instance.setProps({content: 'hello'});

    expect(instance.props.content).not.toBe('hello');
  });

  it('correctly removes stale `aria-expanded` attributes', () => {
    instance = createTippy(h(), {...defaultProps, interactive: true});
    const triggerTarget = h();

    expect(instance.reference.getAttribute('aria-expanded')).toBe('false');

    instance.setProps({interactive: false});

    expect(instance.reference.getAttribute('aria-expanded')).toBe(null);

    instance.setProps({triggerTarget, interactive: true});

    expect(instance.reference.getAttribute('aria-expanded')).toBe(null);
    expect(triggerTarget.getAttribute('aria-expanded')).toBe('false');

    instance.setProps({triggerTarget: null});

    expect(instance.reference.getAttribute('aria-expanded')).toBe('false');
  });
});

describe('instance.setContent()', () => {
  it('works like set({ content: newContent })', () => {
    instance = createTippy(h(), defaultProps);
    const content = 'Hello!';
    instance.setContent(content);

    expect(instance.props.content).toBe(content);
    expect(instance.popperChildren.content.textContent).toBe(content);

    const div = document.createElement('div');
    instance.setContent(div);

    expect(instance.props.content).toBe(div);
    expect(instance.popperChildren.content.firstElementChild).toBe(div);
  });
});

describe('instance.state', () => {
  it('currentPlacement', () => {
    instance = createTippy(h(), {...defaultProps, placement: 'top'});
    instance.show();
    jest.runAllTimers();

    expect(instance.state.currentPlacement).toBe('top');

    instance.setProps({placement: 'bottom'});
    jest.runAllTimers();

    expect(instance.state.currentPlacement).toBe('bottom');

    instance.setProps({placement: 'left-end'});
    jest.runAllTimers();

    expect(instance.state.currentPlacement).toBe('left-end');
  });

  it('isEnabled', () => {
    instance = createTippy(h(), defaultProps);
    instance.show();

    expect(instance.state.isVisible).toBe(true);

    instance.state.isEnabled = false;
    instance.hide();

    expect(instance.state.isVisible).toBe(true);

    instance.state.isEnabled = true;
    instance.hide();

    expect(instance.state.isVisible).toBe(false);

    instance.state.isEnabled = false;
    instance.show();

    expect(instance.state.isVisible).toBe(false);
  });

  it('isVisible', () => {
    instance = createTippy(h(), defaultProps);
    instance.show();

    expect(instance.state.isVisible).toBe(true);

    instance.hide();

    expect(instance.state.isVisible).toBe(false);
  });

  it('isShown', () => {
    instance = createTippy(h(), defaultProps);
    instance.show();

    expect(instance.state.isShown).toBe(false);

    jest.runAllTimers();

    expect(instance.state.isShown).toBe(true);

    instance.hide();

    expect(instance.state.isShown).toBe(false);
  });
});

describe('mountedInstances', () => {
  it('should correctly add and clear instances', () => {
    instance = createTippy(h(), defaultProps);

    instance.show();
    jest.runAllTimers();

    expect(mountedInstances[0]).toBe(instance);

    const instance2 = createTippy(h(), defaultProps);

    instance2.show();
    jest.runAllTimers();

    expect(mountedInstances[0]).toBe(instance);
    expect(mountedInstances[1]).toBe(instance2);

    instance.destroy();

    expect(mountedInstances[0]).toBe(instance2);

    instance2.hide();

    expect(mountedInstances.length).toBe(0);

    instance2.destroy();
  });
});

describe('updateIOSClass', () => {
  beforeEach(enableTouchEnvironment);
  afterEach(disableTouchEnvironment);

  it('should add on mount and remove on unmount', () => {
    instance = createTippy(h(), defaultProps);

    instance.show();
    jest.runAllTimers();

    expect(document.body.classList.contains(IOS_CLASS)).toBe(true);

    instance.hide();

    expect(document.body.classList.contains(IOS_CLASS)).toBe(false);
  });

  it('should only remove if mountedInstances.length is 0', () => {
    instance = createTippy(h(), defaultProps);
    const instance2 = createTippy(h(), defaultProps);

    instance.show();
    instance2.show();
    jest.runAllTimers();

    expect(document.body.classList.contains(IOS_CLASS)).toBe(true);

    instance.hide();

    expect(document.body.classList.contains(IOS_CLASS)).toBe(true);

    instance2.destroy();

    expect(document.body.classList.contains(IOS_CLASS)).toBe(false);
  });
});
