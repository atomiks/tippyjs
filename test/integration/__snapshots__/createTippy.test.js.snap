// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createTippy returns the instance with expected properties 1`] = `
Object {
  "clearDelayTimeouts": [Function],
  "destroy": [Function],
  "disable": [Function],
  "enable": [Function],
  "hide": [Function],
  "id": 1,
  "plugins": Array [],
  "popper": <div
    data-__namespace_prefix__-root=""
    id="__NAMESPACE_PREFIX__-1"
    style="pointer-events: none; z-index: 9999;"
  >
    <div
      class="__NAMESPACE_PREFIX__-box"
      data-animation="fade"
      data-state="hidden"
      role="tooltip"
      style="max-width: 350px;"
      tabindex="-1"
    >
      <div
        class="__NAMESPACE_PREFIX__-content"
        data-state="hidden"
      >
        __DEFAULT_TEST_CONTENT__
      </div>
      <div
        class="__NAMESPACE_PREFIX__-arrow"
      />
    </div>
  </div>,
  "popperInstance": null,
  "props": Object {
    "allowHTML": false,
    "animateFill": false,
    "animation": "fade",
    "appendTo": [Function],
    "aria": Object {
      "content": "describedby",
      "expanded": false,
    },
    "arrow": true,
    "content": "__DEFAULT_TEST_CONTENT__",
    "delay": 0,
    "duration": 0,
    "followCursor": false,
    "getReferenceClientRect": null,
    "hideOnClick": true,
    "ignoreAttributes": false,
    "inertia": false,
    "inlinePositioning": false,
    "interactive": false,
    "interactiveBorder": 2,
    "interactiveDebounce": 0,
    "maxWidth": 350,
    "moveTransition": "",
    "offset": Array [
      0,
      10,
    ],
    "onAfterUpdate": [Function],
    "onBeforeUpdate": [Function],
    "onClickOutside": [Function],
    "onCreate": [Function],
    "onDestroy": [Function],
    "onHidden": [Function],
    "onHide": [Function],
    "onMount": [Function],
    "onShow": [Function],
    "onShown": [Function],
    "onTrigger": [Function],
    "onUntrigger": [Function],
    "placement": "top",
    "plugins": Array [],
    "popperOptions": Object {},
    "render": [Function],
    "role": "tooltip",
    "showOnCreate": false,
    "sticky": false,
    "theme": "",
    "touch": true,
    "trigger": "mouseenter focus",
    "triggerTarget": null,
    "zIndex": 9999,
  },
  "reference": <button
    class="__tippy"
  />,
  "setContent": [Function],
  "setProps": [Function],
  "show": [Function],
  "state": Object {
    "isDestroyed": false,
    "isEnabled": true,
    "isMounted": false,
    "isShown": false,
    "isVisible": false,
  },
  "unmount": [Function],
}
`;
